gsy_framework/matching_algorithms/pay_as_bid_matching_algorithm.py

```diff
@@ -15,33 +15,78 @@ class PayAsBidMatchingAlgorithm(BaseMatchingAlgorithm):
            since the most affordable offers will be allocated for the most aggressive buyers.
     """
 
+    @staticmethod
+    def _match_one_bid_one_offer(offer: Dict, bid: Dict, already_selected_order_energy: Dict,
+                                 market_id: str, time_slot: str) -> Optional[BidOfferMatch]:
+        """
+        Try to match one bid with one offer, and at the same time update the dict with the
+        already selected order energy in order to be able to reuse the same order in future
+        matches.
+        """
+        if (offer.get("energy_rate") - bid.get("energy_rate")) > FLOATING_POINT_TOLERANCE:
+            return None
+
+        if bid["id"] not in already_selected_order_energy:
+            already_selected_order_energy[bid["id"]] = bid["energy"]
+        if offer["id"] not in already_selected_order_energy:
+            already_selected_order_energy[offer["id"]] = offer["energy"]
+        offer_energy = already_selected_order_energy[offer["id"]]
+        bid_energy = already_selected_order_energy[bid["id"]]
+
+        selected_energy = min(offer_energy, bid_energy)
+        if selected_energy <= FLOATING_POINT_TOLERANCE:
+            return None
```

Comment:
In this case I would expect that we need to remove the entries from the `already_selected_order_energy` again, before we return, right?

Reply:
Not really, because absence from the already_selected_order_energy implies that the offer has not been seen yet. This dict works the following way: if there are no entries in it, then we assume that the offer / bid has not iterated over yet. First time that we encounter an offer / bid, we add its energy to this dict, and we gradually decrease this energy while we include this order to matches, until it reaches 0. From then on, we leave it to the dict in order for the loop to know that this order has already depleted its energy.

Reply:
Looking at it again I suppose that it can carry two different meanings

1. already_selected  /  order_energy  # the energy selected per each order (this is what my first impression was)
2. already_selected_order  /  energy  # The orders selected with the associated energy available

Reply:
It basically can be interpreted as either a mapper of the selected energy (energy that was already consumed) or the absolute opposite (the meaning you already explained above)
Anw I can now see your point, if it can be further improved it would be great, otherwise LGTM

Reply:
No worries at all, I intended for the second meaning during implementation, however it was evident from the review that the name was not clear enough, therefore I will rename it to the one you suggested in the other thread, thanks!